ecs-cli/modules/clients/aws/route53/client_test.go

```diff
@@ -0,0 +1,321 @@
+	"github.com/aws/aws-sdk-go/service/route53"
+	"github.com/aws/aws-sdk-go/service/servicediscovery"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestFindPrivateNamespace(t *testing.T) {
+	setupNamespaceMocks(t, false)
+	setupHostedZoneMocks(t)
+
+	var testCases = []struct {
+		testName   string
+		name       string
+		vpc        string
+		region     string
+		expectedID *string
+	}{
+		{
+			testName:   "Find namespace1",
+			name:       "corp",
+			vpc:        "vpc-8BAADF00D",
+			region:     "us-east-1",
+			expectedID: aws.String("namespace1"),
+		},
+		{
+			testName:   "Find namespace2",
+			name:       "prod",
+			vpc:        "vpc-1CEB00DA",
+			region:     "us-east-1",
+			expectedID: aws.String("namespace2"),
+		},
+		{
+			testName:   "Find namespace4",
+			name:       "corp",
+			vpc:        "vpc-C00010FF",
+			region:     "sa-east-1",
+			expectedID: aws.String("namespace4"),
+		},
+		{
+			testName:   "Namespace Does Not Exist",
+			name:       "corp",
+			vpc:        "vpc-C00010FF",
+			region:     "us-east-1",
+			expectedID: nil,
+		},
+		{
+			testName:   "Namspace with multiple VPCs",
+			name:       "bridge",
+			vpc:        "vpc-D15EA5E",
+			region:     "ap-south-1",
+			expectedID: aws.String("namespace3"),
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.testName, func(t *testing.T) {
+			namespaceID, err := FindPrivateNamespace(testCase.name, testCase.vpc, setupParamsInTest(testCase.region))
+			assert.Equal(t, testCase.expectedID, namespaceID, "Expected namespace ID to match")
+			assert.NoError(t, err, "Unexpected error from FindPrivateNamespace")
+		})
+	}
+}
+
+func TestFindPublicNamespace(t *testing.T) {
+	setupNamespaceMocks(t, true)
+
+	var testCases = []struct {
+		testName   string
+		name       string
+		expectedID *string
+	}{
+		{
+			testName:   "Find namespace1",
+			name:       "corp",
+			expectedID: aws.String("namespace1"),
+		},
+		{
+			testName:   "Find namespace2",
+			name:       "prod",
+			expectedID: aws.String("namespace2"),
+		},
+		{
+			testName:   "Find namespace3",
+			name:       "bridge",
+			expectedID: aws.String("namespace3"),
+		},
+		{
+			testName:   "Namespace Does Not Exist",
+			name:       "cat",
+			expectedID: nil,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.testName, func(t *testing.T) {
+			namespaceID, err := FindPublicNamespace(testCase.name, &config.CommandConfig{})
+			assert.Equal(t, testCase.expectedID, namespaceID, "Expected namespace ID to match")
+			assert.NoError(t, err, "Unexpected error from FindPrivateNamespace")
+		})
+	}
+}
+
+// Implements serviceDiscoveryClient interface
+type mockSDClient struct {
+	usingFilter      bool
+	publicNamespaces bool
+	namespaceData    map[string]servicediscovery.Namespace
+	t                *testing.T
+}
+
+func (mock *mockSDClient) ListNamespacesPages(input *servicediscovery.ListNamespacesInput, fn func(*servicediscovery.ListNamespacesOutput, bool) bool) error {
```

Comment:
It feels a little strange to have an implementation of a public API include test assertions this way, but maybe it's just too clever for me...

Reply:
We do this exact thing a lot in our existing tests that use go mock- my code just looks different, but functionally its doing the same thing. See for example: https://github.com/aws/amazon-ecs-cli/blob/master/ecs-cli/modules/cli/cluster/cluster_app_test.go#L168

Any time we use go mock's `.Do()` function on an API call we're adding test assertions on the mock of an API call. 

Also, adding this helped me catch a typo bug which was originally in the code, so this test paid off.